File: Java/Chef.java
// Производный класс Chef наследуется от Employee и реализует множественное наследование
// (от абстрактного класса Employee и интерфейсов Payable, Displayable)
public class Chef extends Employee {
    private int dishesPrepared;
    private String specialty;

    public Chef(String name, double hourlyRate, String specialty, int dishesPrepared) {
        super(name, "Chef", hourlyRate); // Вызов конструктора базового класса
        this.dishesPrepared = dishesPrepared;
        this.specialty = specialty;
    }

    public Chef(String name, double hourlyRate, String specialty) {
        this(name, hourlyRate, specialty, 0);
    }

    // Переопределение метода базового класса с вызовом базового метода
    @Override
    public void showInfo() {
        super.showInfo(); // Вызов метода базового класса
        System.out.println("  Specialty: " + specialty);
        System.out.println("  Dishes Prepared: " + dishesPrepared);
    }

    // Переопределение метода без вызова базового метода
    @Override
    public double calculateSalary() {
        // Повара получают базовую зарплату плюс бонус за блюда
        double baseSalary = hoursWorked * hourlyRate;
        double bonus = dishesPrepared * 10.0; // Бонус 10 за блюдо
        return baseSalary + bonus;
    }

    @Override
    public String getEmployeeType() {
        return "Chef";
    }

    // Новые методы
    public void prepareDish(String dishName) {
        dishesPrepared++;
    }

    public int getDishesPrepared() {
        return dishesPrepared;
    }

    public String getSpecialty() {
        return specialty;
    }

    // Переопределение метода интерфейса Payable
    @Override
    public String getPaymentInfo() {
        return "Chef " + name + " salary: " + calculateSalary() + " (specialty: " + specialty + ")";
    }
}

File: Chef.cpp
#include "Chef.hpp"
#include <iostream>

Chef::Chef(int id, const std::string& name, double rate, const std::string& specialty, int dishesPrepared)
    : Employee(id, name, "Chef", rate), dishesPrepared(dishesPrepared), specialty(specialty)
{
    // Вызов конструктора базового класса Employee в списке инициализации
}

Chef::Chef(const Chef& other)
    : Employee(other), dishesPrepared(other.dishesPrepared), specialty(other.specialty)
{
    // Вызов конструктора копирования базового класса
}

void Chef::showInfo() const {
    // Переопределение метода с вызовом базового метода
    Employee::showInfo(); // Вызов метода базового класса
    std::cout << "  Specialty: " << this->specialty << std::endl;
    std::cout << "  Dishes Prepared: " << this->dishesPrepared << std::endl;
}

double Chef::calculateSalary() const {
    // Переопределение без вызова базового метода
    // Повара получают базовую зарплату плюс бонус за блюда
    double baseSalary = this->hoursWorked * this->hourlyRate;
    double bonus = this->dishesPrepared * 10.0; // Бонус 10 за блюдо
    return baseSalary + bonus;
}

void Chef::prepareDish(const std::string& dishName) {
    this->dishesPrepared++;
}

Chef& Chef::operator=(const Chef& other) {
    if (this != &other) {
        Employee::operator=(other); // Вызов оператора базового класса
        this->dishesPrepared = other.dishesPrepared;
        this->specialty = other.specialty;
    }
    return *this;
}

Chef& Chef::operator=(const Employee& other) {
    Employee::operator=(other); // Присваивание только базовой части
    // Поля, специфичные для Chef, остаются неизменными
    return *this;
}

